src/uiprotect/data/bootstrap.py

```diff
@@ -384,6 +386,58 @@ def _process_remove_packet(
             old_obj=device,
         )
 
+    def _process_ws_keyring_message(
+        self,
+        action: dict[str, Any],
+        data: dict[str, Any],
+    ) -> WSSubscriptionMessage | None:
+        if action["action"] == "add":
+            keyring = create_from_unifi_dict(
+                data, api=self._api, model_type=ModelType.KEYRING
+            )
+            if TYPE_CHECKING:
+                assert isinstance(keyring, Keyring)
+            self.keyrings[keyring.id] = keyring
+            return WSSubscriptionMessage(
+                action=WSAction.ADD,
+                new_update_id=self.last_update_id,
+                changed_data=keyring.dict(),
+                new_obj=keyring,
+            )
+        if action["action"] == "remove":
+            keyring_id = action["id"]
+            keyring = self.keyrings.pop(keyring_id, None)
+            if keyring is None:
+                return None
+            return WSSubscriptionMessage(
+                action=WSAction.REMOVE,
+                new_update_id=self.last_update_id,
+                changed_data={},
+                old_obj=keyring,
+            )
+        if action["action"] == "update":
+            keyring_id = action["id"]
+            keyring = self.keyrings.get(keyring_id)
+            if keyring is None:
+                return None
+            old_keyring = keyring.copy()
+            keyring = keyring.update_from_dict(data)
```

Comment:
```suggestion
            new_keyring = keyring.update_from_dict(data)
```

Reply:
This was a bit confusing so probably better to not reuse the var name or ignore the return value since it should object in place